wremnants/include/muon_calibration.h

```diff
@@ -392,12 +392,79 @@ class calibration_uncertainty_helper {
         return goodGenMuons0_idx_in_recos;
     }
 
+    RVec<bool> filterRecoMuonsByGenTruth(
+        RVec<bool> muonSelection,
+        RVec<int> Muon_genPartIdx,
+        RVec<int> GenPart_pdgId,
+        RVec<int> GenPart_statusFlags,
+        RVec<int> GenPart_status,
+        bool requirePrompt = true
+    ) {
+        ROOT::VecOps::RVec<bool> res(muonSelection.size());
+        for (int i = 0; i < muonSelection.size(); i++) {
+            res[i] = (
+                muonSelection[i] &&
+                (!(requirePrompt) || GenPart_statusFlags[Muon_genPartIdx[i]] & 0x01) &&
+                (abs(GenPart_pdgId[Muon_genPartIdx[i]]) == MUON_PDGID) &&
+                (GenPart_status[Muon_genPartIdx[i]] == 1)
+            );
+        }
+        return res;
+    }
+
+    RVec<bool> filterRecoMuonsByCovMat(
+        RVec<bool> muonSelection,
+        RVec<float> covMat,
+        RVec<int> covMatCounts
+    ) {
+        ROOT::VecOps::RVec<bool> res(muonSelection.size());
+        int covMat_start_idx = 0;
+        for (int i = 0; i < muonSelection.size(); i++) {
+            res[i] = (
+                muonSelection[i] &&
+                //covMatCounts[i] &&
+                covMat[covMat_start_idx] > 0
+            );
+            covMat_start_idx += covMatCounts[i];
+        }
+        return res;
+    }
+
+    template<typename T> RVec<RVec<T>> getCovMatForSelectedRecoMuons(
+        RVec<float> covmat,
+        RVec<int> covmat_counts,
+        RVec<bool> selection,
+        int nMuons = -1 // limit the number of muons to be N; -1 to take all selected muons
+    ) {
+        if (covmat_counts.size() != selection.size()) {
```

Comment:
If this is a fatal error you probably want to throw. Will never see the cout in normal running

Reply:
added throw in latest commit